xdsl/irdl/declarative_assembly_format.py

```diff
@@ -271,79 +275,183 @@ class OptionallyParsableDirective(FormatDirective, ABC):
+      type-directive ::= type(typeable-directive)
+    """
+
+    inner: VariadicTypeableDirective
+
+    def parse_optional(self, parser: Parser, state: ParsingState) -> bool:
+        types = parser.parse_optional_undelimited_comma_separated_list(
+            parser.parse_optional_type, parser.parse_type
+        )
+        if types is None:
+            types = ()
+        self.inner.set_types(types, state)
+        return bool(types)
+
+    def print(self, printer: Printer, state: PrintingState, op: IRDLOperation) -> None:
+        if state.should_emit_space or not state.last_was_punctuation:
+            printer.print(" ")
+        printer.print_list(self.inner.get_types(op), printer.print_attribute)
+        state.last_was_punctuation = False
+        state.should_emit_space = True
+
     def is_present(self, op: IRDLOperation) -> bool:
-        return len(getattr(op, self.name)) > 0
+        return self.inner.is_present(op)
+
+    def set_empty(self, state: ParsingState):
+        self.inner.set_types((), state)
+
+
+@dataclass(frozen=True)
+class OptionalTypeDirective(VariadicLikeTypeDirective):
+    """
+    A directive which parses the type of a optional typeable directive, with format:
+      type-directive ::= type(typeable-directive)
+    """
+
+    inner: OptionalTypeableDirective
+
+    def parse_optional(self, parser: Parser, state: ParsingState) -> bool:
+        type = parser.parse_optional_type()
+        self.inner.set_type(type, state)
+        return bool(type)
 
+    def print(self, printer: Printer, state: PrintingState, op: IRDLOperation) -> None:
+        if state.should_emit_space or not state.last_was_punctuation:
+            printer.print(" ")
+        type = self.inner.get_type(op)
+        if type:
+            printer.print_attribute(type)
+            state.last_was_punctuation = False
+            state.should_emit_space = True
 
-class OptionalVariable(VariadicLikeVariable, ABC):
     def is_present(self, op: IRDLOperation) -> bool:
-        return getattr(op, self.name) is not None
+        return self.inner.is_present(op)
 
+    def set_empty(self, state: ParsingState):
+        self.inner.set_type(None, state)
 
-class VariadicLikeTypeDirective(VariadicLikeFormatDirective, VariableDirective, ABC):
-    pass
+
+@dataclass(frozen=True)
+class VariableDirective(Directive, ABC):
+    """
+    A variable directive, with the following format:
+      variable-directive ::= dollar-ident
+    The directive will request a space to be printed after.
+    """
+
+    name: str
+    """The variable name. This is only used for error message reporting."""
+    index: int
+    """Index of the variable(operand or result) definition."""
 
 
-class VariadicTypeDirective(VariadicLikeTypeDirective, VariadicVariable, ABC):
-    pass
+class VariadicVariable(VariableDirective, AnchorableDirective, ABC):
+    def is_present(self, op: IRDLOperation) -> bool:
+        return len(getattr(op, self.name)) > 0
```

Comment:
```suggestion
        return bool(getattr(op, self.name))
```

Reply:
but then again I wonder whether we should be going via the def instead?

Reply:
We have the index of the variable so should be able to do `op.operands[self.index]`. This should be more type safe too. Should I work that into this PR?

Reply:
You couldn't do the above without more refactoring as `VariadicVariable` doesn't know which type of variable it is

Reply:
Ah too bad, although maybe that would also be worth fixing? I'm happy to punt this side quest until someone stumbles upon it and feels motivated

Reply:
FWIW this is moved code rather than new code (I don't really see this as an excuse though)